Added demandResponse.go file to implement DR listener function

john-b-yang opened the pull request.

gtfierro commented on driver/pelican/demandResponse.go:
```diff
@@ -0,0 +1,152 @@
+}
+
+// Testing function to check whether TrackDREvent function works properly
+func testTrackDREvent() {
```
Can we make this function a comment? Or put it into a separate `test` folder? I'd like the functions in this package to be for the operation of the thermostat and manipulation of its APIs

gtfierro commented on driver/pelican/demandResponse.go:
```diff
@@ -0,0 +1,152 @@
+
+	// Map Status to Corresponding Integer Value
+	statusMode := event.Status
+	if statusMode == "Not Configured" {
```
Can we implement this as a set of constants?

```go
type DR_EVENT_STATUS int
const (
  DR_EVENT_UNKNOWN = -1
  DR_EVENT_NOT_CONFIGURED = iota
  DR_EVENT_UNUSABLE
  DR_EVENT_INACTIVE
// etc...
)
```

Then we can have another function that translates a string to a `DR_EVENT_STATUS`.

gtfierro commented on driver/pelican/demandResponse.go:
```diff
@@ -0,0 +1,152 @@
+
+	// Map Event Type to Corresponding Integer Value
+	typeMode := event.Type
+	if typeMode == "Normal" {
```
Let's implement an enum construction here too. We'll be able to use these enum definitions in other DR-related drivers and applications

jhkolb commented on driver/pelican/demandResponse.go:
```diff
@@ -0,0 +1,152 @@
+}
+
+type ADREvent struct {
+	Event_End   int64
```
I believe the convention in Go is to use camel case, not underscores, so you'll need to change several of the fields in this struct.

jhkolb commented on driver/pelican/demandResponse.go:
```diff
@@ -0,0 +1,152 @@
+		End()
+
+	if errs != nil {
+		return nil, fmt.Errorf("Error retrieving thermostat status from %s: %v", pel.target, errs)
```
Let's make this error message more specific, we're trying to retrieve information about a DR event.

jhkolb commented on driver/pelican/demandResponse.go:
```diff
@@ -0,0 +1,152 @@
+		return nil, fmt.Errorf("Error retrieving thermostat status from %s: %s", resp.Request.URL, result.Message)
+	}
+
+	event := &result.Event
```
Why is the `&` operator used here?

jhkolb commented on driver/pelican/demandResponse.go:
```diff
@@ -0,0 +1,152 @@
+		return nil, fmt.Errorf("Failed to decode response XML: %v", err)
+	}
+	if result.Success == 0 {
+		return nil, fmt.Errorf("Error retrieving thermostat status from %s: %s", resp.Request.URL, result.Message)
```
And let's make this error message clearer too.

jhkolb commented on driver/pelican/demandResponse.go:
```diff
@@ -0,0 +1,152 @@
+
+func DRTimeToUnix(DRTime string, timezone *time.Location) (int64, error) {
+	// Time field is empty or nil
+	if DRTime == "" || len(DRTime) == 0 {
```
Am I confused, or are the two conditions you're checking for redundant? I guess I could imagine some weirdness with non-printable characters, but I would think we only need to check for a length of 0.

jhkolb commented on driver/pelican/demandResponse.go:
```diff
@@ -0,0 +1,152 @@
+		return 0, nil
+	}
+
+	// Using Parse in Location to cast time string into correct int64 value
```
There's no casting going on here, and there's no int64 value yet either.  This will produce a `time.Time` instance with the specified time zone.

jhkolb commented on driver/pelican/demandResponse.go:
```diff
@@ -0,0 +1,152 @@
+)
+
+type ADREventWrapperAPI struct {
+	XMLName xml.Name    `xml:"result"`
```
Hm I don't think I've seen `XMLName` used in the rest of the Pelican code. Why do we need it here? If we don't, we probably want to keep things consistent.

john-b-yang commented:
@jhkolb @gtfierro Thanks for the feedback! I've updated the code to reflect the requested edits. Most significantly, I introduced two sets of an enum and a helper method to better define a DR event's Status and Type. I also removed the test method I originally wrote and folded the logic for publishing the DR status into 'main.go'.

gtfierro commented on driver/pelican/demandResponse.go:
```diff
@@ -43,78 +57,50 @@ func (pel *Pelican) TrackDREvent() (*ADREvent, error) {
 
 	if errs != nil {
-		return nil, fmt.Errorf("Error retrieving thermostat status from %s: %v", pel.target, errs)
+		return nil, fmt.Errorf("Error retrieving thermostat direct-response status from %s: %v", pel.target, errs)
```
`direct-response` should be changed to `demand response` here and elsewhere

gtfierro commented on driver/pelican/demandResponse.go:
```diff
@@ -0,0 +1,142 @@
+	Type   string `xml:"OpenADREventType"`
+}
+
+type ADREvent struct {
```
Can we annotate this struct so that the msgpack serialization is compliant with the [XBOS DR Interface](https://github.com/SoftwareDefinedBuildings/XBOSDocs/blob/master/driver_xbos_demand_response.md)?

Should be able to do this like

```go
type ADREvent struct {
    EventEnd int64 `msgpack:"event_end"`
    // etc
```

gtfierro commented:
Thanks @john-b-yang! This looks 95% good to me. The one last change, which I should have caught earlier (sorry!), is the `i.xbos.demand_response` interface is separate from the `i.xbos.thermostat` interface.

We should have a similar line to https://github.com/john-b-yang/bw2-contrib/blob/ea9f2fd33db9aac9a4676e6e824181a4021b4a69/driver/pelican/main.go#L63 that instantiates a map of `i.xbos.demand_response` interfaces, and then [here](https://github.com/john-b-yang/bw2-contrib/blob/ea9f2fd33db9aac9a4676e6e824181a4021b4a69/driver/pelican/main.go#L168) you should make sure you're publishing from the demand response interface instead of the thermostat interface. The name of the signal should be `info`, not `dr` as well. This is all to just make it compliant with the `i.xbos.demand_response` interface description.

When this is behaving correctly, we should see URIs like 

```
Message from <namespace>/pelican/s.pe  
lican/SouthEastCorner/i.xbos.demand_response/signal/info:  
PO 2.1.1.9 len 62 (msgpack) contents:  
{  
  "dr_status": 3,  
  "event_end": 0,  
  "event_start": 0,  
  "event_type": -1,  
  "time": 1525105880378267371  
}
```

jhkolb commented on driver/pelican/main.go:
```diff
@@ -151,6 +152,22 @@ func main() {
+
+				drStatus, drErr := currentPelican.TrackDREvent()
+				if drErr != nil {
+					fmt.Printf("Failed to retrieve Pelican's DR status: %v\n", err)
```
I think you want `drErr` rather than `err` in this print.

gtfierro commented:
Thanks @john-b-yang ; does the Pelican DR interface expose a timestamp that we're going to have to deal with, like we do with the current thermostat reading?

jhkolb commented:
I don't think the DR API endpoint exposes a timestamp.

Also, do we want to put the publish loops for the two interfaces in separate goroutines? For example, what if we want to poll the thermostat and DR endpoints at different frequencies? This could be overcomplicating things though, so I'm not sure it's really necessary.

gtfierro commented:
Oh, good idea! We'll probably only need to poll the DR interface every hour or so? The pelican thermostat polling is already inside a goroutine so I think that should be a straightforward change. The poll rate for the demand response interface should be another config option.

john-b-yang commented:
@gtfierro @jhkolb I created a separate go routine and poll interval for the demand response polling!

gtfierro commented on driver/pelican/main.go:
```diff
@@ -53,6 +53,13 @@ func main() {
+	pollDrStr := params.MustString("poll_interval_dr")
+	pollDr, drErr := time.ParseDuration(pollDrStr)
+	if drErr != nil {
+		fmt.Printf("Invalid demand response poll interval specified: %v\n", err)
```
Probably put `drErr` here instead of `err`

john-b-yang replied:
Ahhh I keep making these mistakes, I'll be more thorough in the future, sorry about that haha.

gtfierro commented:
I just had one minor fix and then I'm ready to merge this predicated on @jhkolb approval

jhkolb commented on driver/pelican/demandResponse.go:
```diff
@@ -0,0 +1,142 @@
+type DR_EVENT_STATUS int
+
+const (
+	DR_EVENT_UNKNOWN        DR_EVENT_STATUS = -1
```
Strictly speaking, -1 is not a valid value for either the DR event status or type in the current specification of the XBOS interface.

Do we want to remove those values here or update the interface spec to incorporate them?

gtfierro replied:
I like the idea of being able to indicate when something is not understood, but to remain consistent with the other interfaces maybe we want to just drop it when we don't know what's going on (this is the behavior other drivers exhibit). I'm ok with removing these and just printing out an error when we do get an unexpected indicator

john-b-yang replied:
Sounds good! I'll make sure to return an fmt.Error as opposed to assigning an actual error. In that case, I will probably just removed the "unknown" and "unrecognized" cases for the event status and type structs.

jhkolb commented:
This looks ready to me!

jhkolb commented:
I'll leave it to @gtfierro to check that the recent fixes are sufficient and then merge.

gtfierro commented on driver/pelican/main.go:
```diff
@@ -151,9 +162,31 @@ func main() {
+				drStatus, drErr := currentPelican.TrackDREvent()
+				if drErr != nil {
+					fmt.Printf("Failed to retrieve Pelican's DR status: %v\n", drErr)
+					done <- true
```
A problem I noticed when testing is that if we throw an error when parsing the DR event, then the driver exits. This means that we won't be able to run Pelican drivers unless there's a DR event. Maybe we just log the error here and don't cause the DR event to crash the driver? What do folks think?

john-b-yang replied:
@gtfierro @jhkolb That sounds good to me. I can just leave the print statement and remove the "done <- true" statement so driver get request loop does not stop running?

gtfierro replied:
Yes that should work. Just make sure that the driver doesn't then try to publish an empty object. Maybe do the print, then sleep for the poll duration, and then `continue`

gtfierro replied:
Or alternatively set the drStatus to `nil`

jhkolb commented:
Ah, now I see why we had the previous design.

I agree, let's log the fact that we didn't retrieve a valid DR event, but keep the driver running.

gtfierro commented:
Putting this on hold until https://github.com/SoftwareDefinedBuildings/bw2-contrib/pull/25 is fixed

gtfierro commented:
I fixed up the merge and adjusted the behavior in https://github.com/SoftwareDefinedBuildings/bw2-contrib/pull/26
